File: examples/series/invoice-app/invoicing/report.py
"""A small plain-text report: totals per customer and currency.

Amounts in different currencies are never added together; each customer gets
one line per currency. The table is meant for a terminal, so long customer
names are shortened with customers.display_name.
"""
from decimal import Decimal

from .customers import display_name, normalise_name
from .invoice import row_gross
from .money import format_money, to_decimal

NAME_WIDTH = 24


def summarise(rows, currency: str = None) -> list:
    """Aggregate rows into one entry per (customer, currency).

    Each entry is a dict with customer, currency, invoices, net and gross.
    Customers are grouped by their normalised name, so 'Müller GmbH' written
    with a decomposed umlaut still lands on the same line. If ``currency`` is
    given, other currencies are left out.
    """
    wanted = currency.upper() if currency else None
    totals = {}
    for row in rows:
        code = (row.get("currency") or "EUR").upper()
        if wanted and code != wanted:
            continue
        key = (normalise_name(row["customer"]), code)
        entry = totals.setdefault(key, {
            "customer": key[0], "currency": code, "invoices": 0,
            "net": Decimal(0), "gross": Decimal(0),
        })
        entry["invoices"] += 1
        entry["net"] += to_decimal(row["net"])
        entry["gross"] += row_gross(row)
    return [totals[key] for key in sorted(totals)]


def grand_totals(entries) -> dict:
    """Gross total per currency across all customers."""
    result = {}
    for entry in entries:
        result[entry["currency"]] = result.get(entry["currency"], Decimal(0)) + entry["gross"]
    return dict(sorted(result.items()))


def render_table(entries) -> str:
    """Fixed-width text table with a totals footer."""
    header = f"{'Customer':<{NAME_WIDTH}}  {'#':>3}  {'Net':>16}  {'Gross':>16}"
    lines = [header, "-" * len(header)]
    for entry in entries:
        name = display_name(entry["customer"], NAME_WIDTH)
        lines.append(
            f"{name:<{NAME_WIDTH}}  {entry['invoices']:>3}  "
            f"{format_money(entry['net'], entry['currency']):>16}  "
            f"{format_money(entry['gross'], entry['currency']):>16}"
        )
    lines.append("-" * len(header))
    for code, amount in grand_totals(entries).items():
        lines.append(f"{'Total ' + code:<{NAME_WIDTH}}  {'':>3}  {'':>16}  {format_money(amount, code):>16}")
    return "\n".join(lines) + "\n"


def build_report(rows, currency: str = None) -> str:
    """Summarise and render in one call; what the CLI prints."""
    return render_table(summarise(rows, currency=currency))

File: workshops/evals/lab/invoice-app/src/main/java/com/example/invoicing/Report.java
package com.example.invoicing;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.TreeMap;

/**
 * A small plain-text report: totals per customer and currency.
 *
 * <p>Amounts in different currencies are never added together; each customer gets one line
 * per currency. The table is meant for a terminal, so long customer names are shortened with
 * {@link CustomerNames#displayName(String, int)}.
 */
public final class Report {

    static final int NAME_WIDTH = 24;

    private Report() {
    }

    /** One report line: a customer's totals in one currency. */
    public record Entry(String customerName, String currency, int invoices, BigDecimal net, BigDecimal gross) {

        Entry add(Invoice invoice) {
            return new Entry(customerName, currency, invoices + 1, net.add(invoice.net()), gross.add(invoice.gross()));
        }
    }

    /**
     * Aggregates invoices into one entry per (customer, currency). Customers are grouped by
     * their normalised name, so "Müller GmbH" written with a decomposed umlaut lands on the
     * same line. If {@code currency} is not null, other currencies are left out.
     */
    public static List<Entry> summarise(List<Invoice> invoices, String currency) {
        String wanted = currency == null ? null : currency.strip().toUpperCase(Locale.ROOT);
        Map<List<String>, Entry> totals = new LinkedHashMap<>();
        for (Invoice invoice : invoices) {
            if (wanted != null && !invoice.currency().equals(wanted)) {
                continue;
            }
            String customer = CustomerNames.normalise(invoice.customer());
            List<String> key = List.of(customer, invoice.currency());
            Entry entry = totals.getOrDefault(key,
                    new Entry(customer, invoice.currency(), 0, BigDecimal.ZERO, BigDecimal.ZERO));
            totals.put(key, entry.add(invoice));
        }
        List<Entry> entries = new ArrayList<>(totals.values());
        entries.sort(Comparator.comparing(Entry::customerName).thenComparing(Entry::currency));
        return entries;
    }

    /** All currencies. */
    public static List<Entry> summarise(List<Invoice> invoices) {
        return summarise(invoices, null);
    }

    /** Gross total per currency across all customers, ordered by currency code. */
    public static Map<String, BigDecimal> grandTotals(List<Entry> entries) {
        Map<String, BigDecimal> result = new TreeMap<>();
        for (Entry entry : entries) {
            result.merge(entry.currency(), entry.gross(), BigDecimal::add);
        }
        return result;
    }

    /** Fixed-width text table with a totals footer. */
    public static String render(List<Entry> entries) {
        String header = String.format(Locale.ROOT, "%-" + NAME_WIDTH + "s  %3s  %16s  %16s",
                "Customer", "#", "Net", "Gross");
        String rule = "-".repeat(header.length());
        StringBuilder out = new StringBuilder().append(header).append('\n').append(rule).append('\n');
        for (Entry entry : entries) {
            out.append(String.format(Locale.ROOT, "%s  %3d  %16s  %16s\n",
                    pad(CustomerNames.displayName(entry.customerName(), NAME_WIDTH)), entry.invoices(),
                    Money.format(entry.net(), entry.currency()), Money.format(entry.gross(), entry.currency())));
        }
        out.append(rule).append('\n');
        grandTotals(entries).forEach((code, amount) -> out.append(String.format(Locale.ROOT,
                "%-" + NAME_WIDTH + "s  %3s  %16s  %16s\n", "Total " + code, "", "", Money.format(amount, code))));
        return out.toString();
    }

    /** Summarise and render in one call; what the CLI prints. */
    public static String build(List<Invoice> invoices, String currency) {
        return render(summarise(invoices, currency));
    }

    /** Pads to the name column, counting user-perceived characters rather than chars. */
    private static String pad(String name) {
        int length = name.codePointCount(0, name.length());
        return name + " ".repeat(Math.max(0, NAME_WIDTH - length));
    }
}
